### ISSUE TITLE
No macro hygiene for items

### ISSUE THREAD

**Issue by maxsnew** [rust-lang/rust#19700]
Uses of identifiers in a macro expansion refer to identifiers in the macro use-site and not the macro-definition site, effectively giving macros dynamic scope.

Minimal example:
```rust
#![feature(macro_rules)]

fn foo() -> bool {
    true
}

macro_rules! bar(
    () => (foo())
)
    
#[cfg(test)]
mod test {
    fn foo() -> bool {
        false
    }
    #[test]
    fn test() {
        assert!(bar!());
    }
}
```
Testing this program will result in a test failure because when `bar!()` expands to `foo()` the `foo` is the `foo` in the test submodule, whereas it should refer to the `foo` in the top-level module where `bar!` was defined.

This paper is a good resource for implementing this: http://www.cs.utah.edu/plt/publications/macromod.pdf

**Comment by kmcallister** [comment]
This is unfortunate, but it's the [documented behavior](http://doc.rust-lang.org/book/macros.html#hygiene) of `macro_rules!` 1.0 and there is no time to fix it for Rust 1.0. Item/crate hygiene is one of the top priorities for macros two point oh. 

See #22462 for a specific manifestation of this problem that I think we *can* mitigate in time.

**Comment by maxsnew** [comment]
I don't understand why this issue is closed. You're acknowledging it as something that should be fixed so shouldn't it be left open?

**Comment by kmcallister** [comment]
This aspect of `macro_rules!` can't be fixed in a backwards-compatible way, and it can't be fixed in time for Rust 1.0, when the [backwards-compatibility guarantees](http://blog.rust-lang.org/2014/10/30/Stability.html) kick in.

We will at some point introduce a new macro system, first in Rust nightlies, then in some stable Rust 1.x.  This would happen through the [RFC process](https://github.com/rust-lang/rfcs/).  It would coexist with the current `macro_rules!` until Rust 2.0.  The item hygiene issue will not be overlooked by the people designing the new macro system. It is probably the top consideration.

Until there's a full RFC, we can discuss this further on rust-lang/rfcs#440.
